B-Sockets-ServerTesting/sockets-servertesting.md

```diff
@@ -0,0 +1,285 @@
+connection has been made, communication can be done between the server's client
+socket and the client's connecting socket.
+
+
+### HTTP 1.0
+HTTP 1.0 is a protocol between a client, typically a web browser, and a server,
+typically a web server hosting files such as HTML. It is an outdated version of
+the HTTP protocol and simpler than newer versions.
+
+When visiting a website, a URL is specified in the following format:
+```
+ http://example.com:80/index.html
+ ^^^^   ^^^^^^^^^^^ ^^ ^^^^^^^^^^
+ |      |           |  |
+ |      |           |  URI = /index.html
+ |      |           port number = 80
+ |      domain name = example.com
+ protocol = HTTP
+```
+Based on the information provided by the user in the URL, a web client will
+establish a socket connection with the IP address of the domain name. After
+establishing the connection, the two computers exchange text in the form of HTTP
+requests:
+
+- The lient sends a **HTTP request** for a resource on the server
+- The server sends a **HTTP response**
+
+**HTTP request**
+- First line: method, request URI, version
+    - Ex: "GET /index.html HTTP/1.0\r\n"
+- Followed by 0 or more headers
+    - Ex: "Host: www.google.com\r\n"
+- Followed by an empty line
+    - "\r\n"
+
+**HTTP response**
+- First line: response status
+    - Success: "HTTP/1.0 200 OK\r\n"
+    - Failure: "HTTP/1.0 404 Not Found\r\n"
+- Followed by 0 or more response headers
+- Followed by an empty line
+- Followed by the content of the response
+    - Ex: image file or HTML file
+
+We can see examples of these two HTTP requests with `netcat`. Since both
+`netcat` and web servers/clients connect over sockets, we can pretend to be a
+web client or server and recieve the text that the opposing computer sends.
+
+Let's first act as a web server. We first tell `netcat` to open a server
+connection with `nc -l 10000`, and then in a web browser navigate to the URL
+that specifies this server. You can use the domain name `localhost` to specify
+the local computer rather than connecting to a remote computer over the
+internet. In Chrome, I'll navigate to the URL
+`http://localhost:10000/index.html`. `netcat` outputs this:
+```
+$ nc -l 10000
+GET /index.html HTTP/1.1   # GET == method; /index.html == request URI; HTTP/1.0 == version
+Host: localhost:10000      # header
+Connection: keep-alive     # more headers...
+-removed for brevity-
+                           # blank newline to indicate end of headers/request
+```
+
+To act as a client, we can type our HTTP request manually into netcat rather
+than doing it through the web browser. Here, I'll try to send an HTTP request to
+the domain name `example.com` on port `80` (the default for HTTP web servers)
+for the URI `/index.html`. Note that I specify the `-C` with `netcat` so that
+newlines are `\r\n` rather than `\n` -- a requirement of the HTTP protocol. This
+flag may be different depending on your operating system.
```

Comment:
Change "I" to "we" for consistency

Reply:
Flag may be different depending on netcat version (which is related to OS, but can also vary)